crates/ruff_linter/src/rules/ruff/rules/sort_dunder_slots.rs

```diff
@@ -55,74 +75,66 @@ impl Violation for UnsortedDunderSlots {
-/// represented by a `StmtAnnAssign` AST node.
-/// For example: `__slots__: list[str] = ["b", "c", "a"]`.
-pub(crate) fn sort_dunder_slots_ann_assign(checker: &mut Checker, node: &ast::StmtAnnAssign) {
-    if let Some(value) = &node.value {
-        sort_dunder_slots(checker, &node.target, value);
-    }
-}
-
 const SORTING_STYLE: SortingStyle = SortingStyle::Natural;
 
 /// Sort a tuple, list, dict or set that defines `__slots__` in a class scope.
 ///
 /// This routine checks whether the display is sorted, and emits a
 /// violation if it is not sorted. If the tuple/list/set was not sorted,
 /// it attempts to set a `Fix` on the violation.
-fn sort_dunder_slots(checker: &mut Checker, target: &ast::Expr, node: &ast::Expr) {
-    let ast::Expr::Name(ast::ExprName { id, .. }) = target else {
-        return;
+pub(crate) fn sort_dunder_slots(checker: &Checker, binding: &Binding) -> Option<Diagnostic> {
+    let semantic = checker.semantic();
+
+    let (target, value) = match binding.statement(semantic)? {
+        ast::Stmt::Assign(ast::StmtAssign { targets, value, .. }) => match targets.as_slice() {
+            [target] => (target, &**value),
+            _ => return None,
+        },
+        ast::Stmt::AnnAssign(ast::StmtAnnAssign { target, value, .. }) => {
+            (&**target, value.as_deref()?)
+        }
+        _ => return None,
     };
 
+    let ast::ExprName { id, .. } = target.as_name_expr()?;
+
     if id != "__slots__" {
-        return;
+        return None;
     }
 
     // We're only interested in `__slots__` in the class scope
-    let ScopeKind::Class(ast::StmtClassDef {
-        name: class_name, ..
-    }) = checker.semantic().current_scope().kind
-    else {
-        return;
-    };
+    let enclosing_class = semantic.scopes[binding.scope].kind.as_class()?;
 
-    let Some(display) = StringLiteralDisplay::new(node) else {
-        return;
-    };
+    // and it has to be an assignment to a "display literal" (a literal dict/set/tuple/list)
+    let display = StringLiteralDisplay::new(value)?;
 
     let sort_classification = SortClassification::of_elements(&display.elts, SORTING_STYLE);
     if sort_classification.is_not_a_list_of_string_literals() || sort_classification.is_sorted() {
-        return;
+        return None;
     }
 
     let mut diagnostic = Diagnostic::new(
         UnsortedDunderSlots {
-            class_name: class_name.to_string(),
+            class_name: enclosing_class.name.id.clone(),
         },
         display.range,
     );
 
     if let SortClassification::UnsortedAndMaybeFixable { items } = sort_classification {
-        if let Some(fix) = display.generate_fix(&items, checker) {
-            diagnostic.set_fix(fix);
+        if let Some(sorted_source_code) = display.generate_sorted_source_code(&items, checker) {
+            let edit = Edit::range_replacement(sorted_source_code, display.range());
+
+            let applicability = if display.kind.is_set_literal() || !binding.is_used() {
```

Comment:
Nit nit: We may want to consider adding an `is_unused` method. It reads more naturally

Reply:
Agreed -- but there's quite a few other places where we might consider using that method, so I'll propose it in a followup

Reply:
See https://github.com/astral-sh/ruff/pull/12729